File: automated_scraper.py
#!/usr/bin/env python3
"""
Automated Rankings Scraper for Justin Boone Fantasy Football Rankings

This script automates the scraping of all position rankings from Yahoo Sports/Datawrapper
and generates the rankings.json file automatically.
"""

from playwright.sync_api import sync_playwright
import pandas as pd
import json
import time
import logging
from typing import Dict, List, Optional
from dataclasses import dataclass
from datetime import datetime

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

@dataclass
class ScrapingConfig:
    """Configuration for scraping different positions"""
    position: str
    name: str
    urls: List[str]  # Multiple URLs to try if one fails
    datawrapper_selector: str = 'table'
    
# Configuration for all positions - UPDATE THESE URLs AS NEEDED
SCRAPING_CONFIGS = [
    ScrapingConfig(
        position="qb",
        name="Quarterbacks",
        urls=[
            "https://datawrapper.dwcdn.net/bqgx9/",  # Your current QB URL
            # Add backup URLs if available
        ]
    ),
    ScrapingConfig(
        position="rb", 
        name="Running Backs",
        urls=[
            # Add RB Datawrapper URLs here
            "https://datawrapper.dwcdn.net/REPLACE_WITH_RB_ID/",
        ]
    ),
    ScrapingConfig(
        position="wr",
        name="Wide Receivers", 
        urls=[
            # Add WR Datawrapper URLs here
            "https://datawrapper.dwcdn.net/REPLACE_WITH_WR_ID/",
        ]
    ),
    ScrapingConfig(
        position="te",
        name="Tight Ends",
        urls=[
            # Add TE Datawrapper URLs here
            "https://datawrapper.dwcdn.net/REPLACE_WITH_TE_ID/",
        ]
    ),
    ScrapingConfig(
        position="flex",
        name="FLEX",
        urls=[
            # Add FLEX Datawrapper URLs here
            "https://datawrapper.dwcdn.net/REPLACE_WITH_FLEX_ID/",
        ]
    ),
    ScrapingConfig(
        position="def",
        name="Defense/ST",
        urls=[
            # Add DEF Datawrapper URLs here  
            "https://datawrapper.dwcdn.net/REPLACE_WITH_DEF_ID/",
        ]
    ),
    ScrapingConfig(
        position="k",
        name="Kickers",
        urls=[
            # Add K Datawrapper URLs here
            "https://datawrapper.dwcdn.net/REPLACE_WITH_K_ID/",
        ]
    )
]

class RankingsScraper:
    """Main scraper class that handles all positions"""
    
    def __init__(self, headless: bool = True, timeout: int = 30000):
        self.headless = headless
        self.timeout = timeout
        self.scraped_data = {}
        
    def scrape_position_datawrapper(self, config: ScrapingConfig) -> List[Dict]:
        """Scrape rankings for a specific position from Datawrapper iframes"""
        logger.info(f"Starting scrape for {config.name} ({config.position})")
        
        for url in config.urls:
            try:
                if "REPLACE_WITH" in url:
                    logger.warning(f"Skipping {config.position} - URL not configured: {url}")
                    continue
                    
                logger.info(f"Attempting to scrape {config.position} from: {url}")
                rankings = self._scrape_datawrapper_url(url, config)
                
                if rankings:
                    logger.info(f"Successfully scraped {len(rankings)} {config.name}")
                    return rankings
                else:
                    logger.warning(f"No data found for {config.position} from {url}")
                    
            except Exception as e:
                logger.error(f"Error scraping {config.position} from {url}: {e}")
                continue
                
        logger.error(f"Failed to scrape {config.name} from all configured URLs")
        return []
    
    def _scrape_datawrapper_url(self, url: str, config: ScrapingConfig) -> List[Dict]:
        """Scrape a single Datawrapper URL"""
        with sync_playwright() as p:
            browser = p.chromium.launch(headless=self.headless)
            page = browser.new_page()
            
            try:
                logger.info(f"Loading {url}...")
                response = page.goto(url, timeout=self.timeout)
                logger.info(f"Response status: {response.status}")
                
                # Wait for table to load
                page.wait_for_selector('table', timeout=15000)
                time.sleep(3)  # Additional wait for dynamic content
                
                rankings = []
                rows = page.locator('tbody tr[class*="svelte"]').all()
                logger.info(f"Found {len(rows)} data rows")
                
                for i, row in enumerate(rows):
                    try:
                        ranking_data = self._extract_row_data(row, config.position)
                        if ranking_data:
                            rankings.append(ranking_data)
                            logger.debug(f"  {ranking_data['preGameRank']}. {ranking_data['player']} {ranking_data['opponent']}")
                    
                    except Exception as e:
                        logger.warning(f"Error processing row {i}: {e}")
                        continue
                
                return rankings
                
            except Exception as e:
                logger.error(f"Error scraping {url}: {e}")
                return []
                
            finally:
                browser.close()
    
    def _extract_row_data(self, row, position: str) -> Optional[Dict]:
        """Extract data from a single table row"""
        try:
            # Get rank from th element
            rank_elem = row.locator('th[class*="svelte"]').first
            rank = rank_elem.inner_text().strip() if rank_elem.count() > 0 else ""
            
            # Get all td elements
            td_elements = row.locator('td[class*="svelte"]').all()
            
            if len(td_elements) < 2:
                return None
                
            player = td_elements[0].inner_text().strip()
            opponent = td_elements[-1].inner_text().strip() if len(td_elements) > 1 else ""
            
            # Extract team info if available
            team = ""
            if len(td_elements) >= 3:
                team_elem = td_elements[1]
                img = team_elem.locator('img').first
                if img.count() > 0:
                    img_src = img.get_attribute('src')
                    team = self._extract_team_from_url(img_src)
            
            # Only return valid data
            if rank and player and rank.isdigit():
                ranking_data = {
                    'preGameRank': int(rank),
                    'player': player,
                    'opponent': opponent
                }
                
                # Add position rank for FLEX
                if position == "flex" and team:
                    ranking_data['positionRank'] = f"{self._guess_position(player)}{rank}"
                
                return ranking_data
                
        except Exception as e:
            logger.warning(f"Error extracting row data: {e}")
            
        return None
    
    def _extract_team_from_url(self, img_src: str) -> str:
        """Extract team abbreviation from image URL"""
        if not img_src:
            return ""
            
        url_parts = img_src.split('/')
        for part in url_parts:
            if '.png' in part or '.webp' in part:
                team_part = part.split('.')[0]
                if '_' in team_part:
                    return team_part.split('_')[-1].upper()
                elif '2019_' in part:
                    return part.split('2019_')[-1].split('.')[0].upper()
        return ""
    
    def _guess_position(self, player_name: str) -> str:
        """Guess position based on player name - basic implementation"""
        # This is a simplified approach - you might want to maintain a player database
        # or use more sophisticated logic
        common_rb_names = ['McCaffrey', 'Henry', 'Cook', 'Kamara', 'Jones', 'Taylor']
        common_wr_names = ['Chase', 'Jefferson', 'Hill', 'Adams', 'Hopkins', 'Evans']
        
        for rb_name in common_rb_names:
            if rb_name in player_name:
                return "RB"
                
        for wr_name in common_wr_names:
            if wr_name in player_name:
                return "WR"
                
        return "RB"  # Default assumption
    
    def scrape_all_positions(self) -> Dict:
        """Scrape all configured positions"""
        logger.info("Starting automated scraping for all positions...")
        results = {}
        
        for config in SCRAPING_CONFIGS:
            rankings = self.scrape_position_datawrapper(config)
            results[config.position] = rankings
            
            if rankings:
                logger.info(f"✅ {config.name}: {len(rankings)} players")
            else:
                logger.warning(f"❌ {config.name}: No data scraped")
                
            # Brief pause between positions
            time.sleep(2)
        
        self.scraped_data = results
        return results
    
    def generate_rankings_json(self, output_file: str = "rankings.json") -> bool:
        """Generate the rankings.json file from scraped data"""
        try:
            if not self.scraped_data:
                logger.error("No scraped data available. Run scrape_all_positions() first.")
                return False
            
            # Create backup of existing file
            try:
                with open(output_file, 'r') as f:
                    existing_data = f.read()
                backup_file = f"{output_file}.backup.{datetime.now().strftime('%Y%m%d_%H%M%S')}"
                with open(backup_file, 'w') as f:
                    f.write(existing_data)
                logger.info(f"Created backup: {backup_file}")
            except FileNotFoundError:
                logger.info("No existing rankings.json file to backup")
            
            # Write new rankings
            with open(output_file, 'w') as f:
                json.dump(self.scraped_data, f, indent=2)
                
            logger.info(f"✅ Successfully generated {output_file}")
            
            # Print summary
            self._print_summary()
            return True
            
        except Exception as e:
            logger.error(f"Error generating rankings JSON: {e}")
            return False
    
    def _print_summary(self):
        """Print a summary of scraped data"""
        print("\n" + "="*50)
        print("SCRAPING SUMMARY")
        print("="*50)
        
        total_players = 0
        for position, rankings in self.scraped_data.items():
            count = len(rankings)
            total_players += count
            status = "✅" if count > 0 else "❌"
            print(f"{status} {position.upper()}: {count} players")
        
        print(f"\nTotal players scraped: {total_players}")
        print("="*50)

def main():
    """Main execution function"""
    print("🏈 Justin Boone Rankings Automated Scraper")
    print("=" * 50)
    
    # Create scraper instance
    scraper = RankingsScraper(headless=False)  # Set to True for production
    
    try:
        # Scrape all positions
        results = scraper.scrape_all_positions()
        
        # Generate rankings.json
        success = scraper.generate_rankings_json()
        
        if success:
            print("\n🎉 Automated scraping completed successfully!")
            print("📄 rankings.json has been updated with the latest data")
            print("\nNext steps:")
            print("1. Review the generated rankings.json file")
            print("2. Commit and push your changes to deploy")
            print("3. Update the environment variable LAST_STORED_TIMESTAMP")
        else:
            print("\n❌ Failed to generate rankings.json file")
            
    except KeyboardInterrupt:
        print("\n⚠️ Scraping interrupted by user")
    except Exception as e:
        logger.error(f"Unexpected error in main: {e}")
        print(f"\n❌ Scraping failed: {e}")

if __name__ == "__main__":
    main()

File: netlify/functions/automated-scraper.js
// Netlify function for automated Boone rankings scraping
const { chromium } = require('playwright-chromium');
const fetch = require('node-fetch');
const cheerio = require('cheerio');

// Configuration - you can also use environment variables
const DEFAULT_CONFIG = {
    monitor_url: process.env.MONITOR_URL || "",
    datawrapper_urls: {
        qb: (process.env.QB_URLS || "").split(',').filter(url => url.trim()),
        rb: (process.env.RB_URLS || "").split(',').filter(url => url.trim()),
        wr: (process.env.WR_URLS || "").split(',').filter(url => url.trim()),
        te: (process.env.TE_URLS || "").split(',').filter(url => url.trim()),
        flex: (process.env.FLEX_URLS || "").split(',').filter(url => url.trim()),
        def: (process.env.DEF_URLS || "").split(',').filter(url => url.trim()),
        k: (process.env.K_URLS || "").split(',').filter(url => url.trim())
    },
    last_stored_timestamp: process.env.LAST_STORED_TIMESTAMP || null
};

exports.handler = async (event, context) => {
    console.log('🤖 Starting automated Boone rankings check...');
    
    try {
        // Check if we're in active hours (6 AM ET to 6 PM ET)
        const now = new Date();
        const etHour = getETHour(now);
        const isActiveHours = etHour >= 6 && etHour < 18;
        
        if (!isActiveHours) {
            console.log(`⏰ Outside active hours (${etHour}:00 ET). Boone only updates between 6 AM and 6 PM ET.`);
            return {
                statusCode: 200,
                body: JSON.stringify({
                    success: true,
                    message: `Outside active hours (${etHour}:00 ET). Next check will be during 6 AM - 6 PM ET.`
                })
            };
        }
        
        console.log(`✅ Active hours (${etHour}:00 ET) - proceeding with check`);
        
        // Get configuration (prefer environment variables, fallback to hardcoded)
        const config = await getConfig();
        
        if (!config.monitor_url) {
            console.log('❌ No monitor URL configured');
            return {
                statusCode: 200,
                body: JSON.stringify({
                    success: true,
                    message: 'No monitor URL configured'
                })
            };
        }
        
        // Check for timestamp changes
        const timestampResult = await checkTimestampChange(config.monitor_url, config.last_stored_timestamp);
        
        if (!timestampResult.updateNeeded) {
            console.log('✅ No update needed - timestamps match');
            return {
                statusCode: 200,
                body: JSON.stringify({
                    success: true,
                    updateNeeded: false,
                    message: 'No update needed',
                    currentTimestamp: timestampResult.currentTimestamp
                })
            };
        }
        
        console.log('🚨 Update detected! Starting automated scraping...');
        
        // Scrape all configured positions
        const scrapingResults = await scrapeAllPositions(config.datawrapper_urls);
        
        // Check if we got any data
        const totalPlayers = Object.values(scrapingResults).reduce((sum, rankings) => sum + rankings.length, 0);
        
        if (totalPlayers === 0) {
            console.log('❌ No data was scraped from any position');
            return {
                statusCode: 500,
                body: JSON.stringify({
                    success: false,
                    error: 'No data scraped from any position'
                })
            };
        }
        
        console.log(`✅ Scraped ${totalPlayers} total players`);
        
        // Log the results (in a real deployment, you'd save to a database or file storage)
        console.log('📊 Scraping Summary:');
        Object.entries(scrapingResults).forEach(([position, rankings]) => {
            console.log(`  ${position.upper()}: ${rankings.length} players`);
        });
        
        // In a real implementation, you'd:
        // 1. Save scrapingResults to your database/file storage
        // 2. Update the LAST_STORED_TIMESTAMP environment variable
        // 3. Trigger your site rebuild
        
        console.log('🎉 Automated scraping completed successfully!');
        
        return {
            statusCode: 200,
            body: JSON.stringify({
                success: true,
                updateNeeded: true,
                totalPlayers: totalPlayers,
                results: scrapingResults,
                currentTimestamp: timestampResult.currentTimestamp,
                message: 'Automated scraping completed successfully'
            })
        };
        
    } catch (error) {
        console.error('❌ Error in automated scraper:', error);
        
        return {
            statusCode: 500,
            body: JSON.stringify({
                success: false,
                error: error.message
            })
        };
    }
};

async function getConfig() {
    // In a real implementation, you might load this from a database or file storage
    // For now, we'll use environment variables with fallbacks
    
    const config = {
        monitor_url: process.env.MONITOR_URL || DEFAULT_CONFIG.monitor_url,
        last_stored_timestamp: process.env.LAST_STORED_TIMESTAMP || DEFAULT_CONFIG.last_stored_timestamp,
        datawrapper_urls: {}
    };
    
    // Load Datawrapper URLs from environment variables
    const positions = ['qb', 'rb', 'wr', 'te', 'flex', 'def', 'k'];
    positions.forEach(position => {
        const envKey = `${position.toUpperCase()}_URLS`;
        const envValue = process.env[envKey] || "";
        config.datawrapper_urls[position] = envValue.split(',').map(url => url.trim()).filter(url => url);
    });
    
    console.log('📋 Loaded configuration:');
    console.log(`  Monitor URL: ${config.monitor_url ? '✅' : '❌'}`);
    console.log(`  Last timestamp: ${config.last_stored_timestamp || 'None'}`);
    
    positions.forEach(position => {
        const count = config.datawrapper_urls[position].length;
        console.log(`  ${position.toUpperCase()}: ${count} URL(s)`);
    });
    
    return config;
}

async function checkTimestampChange(monitorUrl, lastStoredTimestamp) {
    try {
        console.log(`🔍 Checking timestamp at: ${monitorUrl}`);
        
        const response = await fetch(monitorUrl, {
            headers: {
                'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
            }
        });
        
        if (!response.ok) {
            throw new Error(`HTTP ${response.status} - ${response.statusText}`);
        }
        
        const html = await response.text();
        const $ = cheerio.load(html);
        
        // Extract timestamp from content-timestamp div
        const timestampElement = $('.content-timestamp time');
        const currentTimestamp = timestampElement.attr('datetime');
        
        if (!currentTimestamp) {
            throw new Error('No timestamp found on page - page structure may have changed');
        }
        
        const updateNeeded = !lastStoredTimestamp || currentTimestamp !== lastStoredTimestamp;
        
        console.log(`📊 Timestamp comparison:`);
        console.log(`    Current: ${currentTimestamp}`);
        console.log(`    Last stored: ${lastStoredTimestamp || 'None'}`);
        console.log(`    Update needed: ${updateNeeded ? 'YES' : 'NO'}`);
        
        return {
            updateNeeded,
            currentTimestamp,
            lastStoredTimestamp
        };
        
    } catch (error) {
        console.error('Error checking timestamp:', error);
        throw error;
    }
}

async function scrapeAllPositions(datawrapperUrls) {
    const results = {};
    
    // Launch browser
    const browser = await chromium.launch();
    
    try {
        for (const [position, urls] of Object.entries(datawrapperUrls)) {
            if (!urls || urls.length === 0) {
                results[position] = [];
                continue;
            }
            
            console.log(`🔍 Scraping ${position.toUpperCase()} from ${urls.length} URL(s)`);
            
            const rankings = await scrapePosition(browser, position, urls);
            results[position] = rankings;
            
            console.log(`  ${position.toUpperCase()}: ${rankings.length} players`);
            
            // Brief pause between positions
            await new Promise(resolve => setTimeout(resolve, 2000));
        }
        
        return results;
        
    } finally {
        await browser.close();
    }
}

async function scrapePosition(browser, position, urls) {
    for (const url of urls) {
        try {
            console.log(`  Attempting ${position.toUpperCase()}: ${url}`);
            
            const page = await browser.newPage();
            
            try {
                await page.goto(url, { timeout: 30000 });
                await page.waitForSelector('table', { timeout: 15000 });
                await page.waitForTimeout(3000); // Wait for dynamic content
                
                const rankings = [];
                
                const rows = await page.locator('tbody tr[class*="svelte"]').all();
                
                for (const row of rows) {
                    try {
                        const rankingData = await extractRowData(row, position);
                        if (rankingData) {
                            rankings.push(rankingData);
                        }
                    } catch (e) {
                        console.log(`    Warning: Error processing row: ${e.message}`);
                    }
                }
                
                await page.close();
                
                if (rankings.length > 0) {
                    console.log(`    ✅ Successfully scraped ${rankings.length} players`);
                    return rankings;
                } else {
                    console.log(`    ❌ No data found`);
                }
                
            } catch (e) {
                await page.close();
                throw e;
            }
            
        } catch (error) {
            console.log(`    ❌ Error with URL: ${error.message}`);
        }
    }
    
    console.log(`  ❌ Failed to scrape ${position.toUpperCase()} from all URLs`);
    return [];
}

async function extractRowData(row, position) {
    try {
        // Get rank from th element
        const rankElem = row.locator('th[class*="svelte"]').first();
        const rank = await rankElem.innerText();
        const rankNum = rank.trim();
        
        // Get all td elements
        const tdElements = await row.locator('td[class*="svelte"]').all();
        
        if (tdElements.length < 2 || !rankNum || !rankNum.match(/^\d+$/)) {
            return null;
        }
        
        const player = await tdElements[0].innerText();
        const opponent = await tdElements[tdElements.length - 1].innerText();
        
        const rankingData = {
            preGameRank: parseInt(rankNum),
            player: player.trim(),
            opponent: opponent.trim()
        };
        
        // Add position rank for FLEX
        if (position === 'flex') {
            const guessedPosition = guessPosition(player.trim());
            rankingData.positionRank = `${guessedPosition}${rankNum}`;
        }
        
        return rankingData;
        
    } catch (error) {
        return null;
    }
}

function guessPosition(playerName) {
    const rbKeywords = ['McCaffrey', 'Henry', 'Cook', 'Kamara', 'Barkley', 'Robinson', 'Gibbs'];
    const wrKeywords = ['Chase', 'Jefferson', 'Hill', 'Adams', 'Evans', 'Brown', 'Lamb'];
    const teKeywords = ['Kelce', 'Andrews', 'Kittle', 'Waller', 'Hockenson', 'LaPorta'];
    
    const playerUpper = playerName.toUpperCase();
    
    for (const keyword of teKeywords) {
        if (playerUpper.includes(keyword.toUpperCase())) {
            return 'TE';
        }
    }
    
    for (const keyword of wrKeywords) {
        if (playerUpper.includes(keyword.toUpperCase())) {
            return 'WR';
        }
    }
    
    for (const keyword of rbKeywords) {
        if (playerUpper.includes(keyword.toUpperCase())) {
            return 'RB';
        }
    }
    
    return 'RB'; // Default
}

function getETHour(date) {
    // Convert current time to Eastern Time
    const utcHour = date.getUTCHours();
    const month = date.getUTCMonth() + 1;
    const day = date.getUTCDate();
    
    // Simple DST calculation for Eastern Time
    const isDST = (month > 3 && month < 11) || 
                  (month === 3 && day >= 8) || 
                  (month === 11 && day <= 7);
    
    const etOffset = isDST ? -4 : -5; // EDT (-4) or EST (-5)
    let etHour = utcHour + etOffset;
    
    // Handle day rollover
    if (etHour < 0) etHour += 24;
    if (etHour >= 24) etHour -= 24;
    
    return etHour;
}